### PR TITLE
Add dpctl tensor support

### PR DESCRIPTION
**Pull request by ZzEeKkAa**
Add support for `dpctl.tensor.usm_ndarray` to dpjit and experimental kernel.

Several additional changes where made:
- Apply better naming for `USMNdArray`/`DpnpNdArray` data models.
- Fix for `USMNdArray` device data model.
- Use same data type for host and separate type for device for both `USMNdArray` and `DpnpNdArray`

Checklist
- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: numba_dpex/core/datamodel/models.py [hunk 1]
```diff
@@ -55,58 +55,68 @@ def __init__(self, dmm, fe_type):
...
+    members in the array.
+
+    We need both HostModel and DeviceModel to provide address space attribute
+    for some old drivers. Another point is that we don't need some of the fields
+    on device. Since they are flattened on host and reconstructed on device we
+    can save some time for it. All parent and memory management happens on host.
+
+    Read more about it in _init_data_model_manager() doc.
+
+    As been said above we don't need meminfo, parent and sycl_queue, since those
```
Comment (diptorupd) [inline numba_dpex/core/datamodel/models.py:72]: Can be removed, look at my edits to the first paragraph of this docstring.

File: numba_dpex/core/datamodel/models.py [hunk 2]
```diff
@@ -55,58 +55,68 @@ def __init__(self, dmm, fe_type):
...
+class USMArrayDeviceModel(StructModel):
+    """A data model to represent a usm array types in LLVM IR on device.
 
     Dpex's ArrayModel is based on Numba's ArrayModel for NumPy arrays. The
-    dpex model adds an extra address space attribute to all pointer members
-    in the array.
+    dpex model adds an extra address space attribute to almost all pointer
+    members in the array.
+
+    We need both HostModel and DeviceModel to provide address space attribute
```
Comment (diptorupd) [inline numba_dpex/core/datamodel/models.py:65]: The paragraph can be removed. Look at my earlier proposed edit.

File: numba_dpex/core/datamodel/models.py [hunk 3]
```diff
@@ -55,58 +55,68 @@ def __init__(self, dmm, fe_type):
...
 
-class USMArrayModel(StructModel):
-    """A data model to represent a Dpex's array types in LLVM IR.
+class USMArrayDeviceModel(StructModel):
+    """A data model to represent a usm array types in LLVM IR on device.
 
     Dpex's ArrayModel is based on Numba's ArrayModel for NumPy arrays. The
-    dpex model adds an extra address space attribute to all pointer members
-    in the array.
+    dpex model adds an extra address space attribute to almost all pointer
```
Comment (diptorupd) [inline numba_dpex/core/datamodel/models.py:62]: The first two paragraphs in the docstring may be updated as follows. Basically, let us be explicit with the details.

>The ``USMArrayDeviceModel`` adds an extra address space attribute to the ``data`` member. The extra attribute is needed when passing usm_ndarray array arguments to kernels that are compiled for certain OpenCL GPU devices. Note that the address space attribute is applied only to the ``data`` member and not other members of ``USMArrayDeviceModel`` that are pointers. It is done this way as other pointer members such as ``meminfo`` are not used inside a kernel and these members maybe removed from the ``USMArrayDeviceModel`` in future (refer #929).
>
> We use separate data models for host (``USMArrayHostModel``) and device (``USMArrayDeviceModel``) as the address space attribute is only required for kernel functions and not needed for functions that are compiled for a host memory space.

File: numba_dpex/core/datamodel/models.py [hunk 4]
```diff
@@ -55,58 +55,68 @@ def __init__(self, dmm, fe_type):
...
+        """
+        Return the number of fields in an instance of a USMArrayDeviceModel.
+        """
         return _get_flattened_member_count(self)
 
 
-class DpnpNdArrayModel(StructModel):
-    """Data model for the DpnpNdArray type.
+class USMArrayHostModel(StructModel):
+    """Data model for the USMNdArray type.
```
Comment (diptorupd) [inline numba_dpex/core/datamodel/models.py:109]:
```suggestion
    """Data model for the USMNdArray type when used in a host-only function.
```

File: numba_dpex/core/datamodel/models.py [hunk 5]
```diff
@@ -55,58 +55,68 @@ def __init__(self, dmm, fe_type):
         super(GenericPointerModel, self).__init__(dmm, fe_type, be_type)
 
 
-class USMArrayModel(StructModel):
-    """A data model to represent a Dpex's array types in LLVM IR.
+class USMArrayDeviceModel(StructModel):
+    """A data model to represent a usm array types in LLVM IR on device.
```
Comment (diptorupd) [inline numba_dpex/core/datamodel/models.py:59]:
```suggestion
    """A data model to represent a usm array type in the LLVM IR generated for a device-only kernel function.
```

File: numba_dpex/core/datamodel/models.py [hunk 6]
```diff
@@ -242,28 +252,28 @@ def _init_data_model_manager() -> datamodel.DataModelManager:
...
     pointer arguments that have no address space qualifier causes a run time
     crash. For this reason, numba-dpex defines two separate data
-    models: USMArrayModel and DpnpNdArrayModel. When a dpnp.ndarray object is
-    passed as an argument to a ``numba_dpex.kernel`` decorated function it uses
-    the USMArrayModel and when passed to a ``numba_dpex.dpjit`` decorated
-    function it uses the DpnpNdArrayModel. The difference is due to the fact
-    that inside a ``dpjit`` decorated function a dpnp.ndarray object can be
-    passed to any other regular function.
+    models: USMArrayDeviceModel and USMArrayHostModel. When a dpnp.ndarray
+    object is passed as an argument to a ``numba_dpex.kernel`` decorated\
```
Comment (diptorupd) [inline numba_dpex/core/datamodel/models.py:256]: Why the `\`?

File: numba_dpex/core/types/dpnp_ndarray_type.py [hunk 7]
```diff
@@ -200,19 +200,22 @@ def __allocate__(
...
 
+# TODO: move this section to separate file
 # --------------- Boxing/Unboxing logic for dpnp.ndarray ----------------------#
 
 
-@unbox(DpnpNdArray)
+@unbox(USMNdArray)
 def unbox_dpnp_nd_array(typ, obj, c):
-    """Converts a dpnp.ndarray object to a Numba internal array structure.
+    """Converts a dpctl.tensor.usm_ndarray/dpnp.ndarray object to a Numba
```
Comment (diptorupd) [inline numba_dpex/core/types/dpnp_ndarray_type.py:209]:
```suggestion
    """Converts a dpctl.tensor.usm_ndarray/dpnp.ndarray object to a Numba-dpex
```

File: numba_dpex/core/types/dpnp_ndarray_type.py [hunk 8]
```diff
@@ -264,24 +267,26 @@ def unbox_dpnp_nd_array(typ, obj, c):
...
-        typ: The representation of the DpnpNdArray type.
-        val: A native representation of a Numba DpnpNdArray type object.
+        typ: The representation of the USMNdArray/DpnpNdArray type.
+        val: A native representation of a Numba USMNdArray/DpnpNdArray type
+            object.
         c: The boxing context.
 
-    Returns: A Pyobject for a dpnp.ndarray boxed from the Numba native value.
+    Returns: A Pyobject for a dpctl.tensor.usm_ndarray/dpnp.ndarray boxed from
+        the Numba native value.
```
Comment (diptorupd) [inline numba_dpex/core/types/dpnp_ndarray_type.py:289]:
```suggestion
        the Numba-dpex native value.
```

File: numba_dpex/tests/experimental/test_inputs.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
+import dpctl.tensor as dpt
```
Comment (diptorupd) [inline numba_dpex/tests/experimental/test_inputs.py:1]: The test case should be renamed to `test_supported_array_types_as_kernel_args.py`. Verbose but captures what the test is checking.

File: numba_dpex/core/datamodel/models.py (hunk 6 above)
Comment (ZzEeKkAa) [reply numba_dpex/core/datamodel/models.py:256]: mistype

**Review by diptorupd — APPROVED** [review]
Thank you!

**Pull request merged**